File: 백준/Gold/1753. 최단경로/최단경로.py
from queue import PriorityQueue
import sys
read = sys.stdin.readline

n, e = map(int, read().split())
s = int(read())
l = [[] for _ in range(n + 1)]

for i in range(e):
    u, v, w = map(int, read().split())
    l[u].append((v, w))

distance = [sys.maxsize] * (n + 1)
distance[s] = 0
visited = [False] * (n + 1)

queue = PriorityQueue()
queue.put((distance[s], s))

while not queue.empty():
    p = queue.get()
    if visited[p[1]]:
        continue
    visited[p[1]] = True
    for i in l[p[1]]:
        if distance[i[0]] > distance[p[1]] + i[1]:
            distance[i[0]] = distance[p[1]] + i[1]
            queue.put((distance[i[0]], i[0]))

for i in range(1, n + 1):
    if visited[i]:
        print(distance[i])
    else:
        print('INF')

File: 백준/Gold/1753. 최단경로/최단경로.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

class Node implements Comparable<Node>{
	int target, weight;
	
	Node(int target, int weight){
		this.target = target;
		this.weight = weight;
	}

	@Override
	public int compareTo(Node o) {
		return this.weight - o.weight;
	}
}

public class Main {

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		StringBuilder sb = new StringBuilder();
		
		int v = Integer.parseInt(st.nextToken());
		int e = Integer.parseInt(st.nextToken());
		
		int k = Integer.parseInt(br.readLine());
		
		ArrayList<Node>[] graph = new ArrayList[v + 1];
		
		for(int i = 1; i <= v; i++)
			graph[i] = new ArrayList<>();
			
		for(int i = 0; i < e; i++) {
			st = new StringTokenizer(br.readLine());
			int startNode = Integer.parseInt(st.nextToken());
			int endNode = Integer.parseInt(st.nextToken());
			int value = Integer.parseInt(st.nextToken());
			
			graph[startNode].add(new Node(endNode, value));
		}
		
		int[] dist = new int[v + 1];
		boolean[] visited = new boolean[v + 1];
		
		for(int i = 1; i <= v; i++)
			dist[i] = Integer.MAX_VALUE;
		
		PriorityQueue<Node> pq = new PriorityQueue<>();
		pq.add(new Node(k, 0));
		dist[k] = 0;
		
		while(!pq.isEmpty()) {
			Node node = pq.poll();
			int target = node.target;
			
			if(!visited[target]) {
				visited[target] = true;
				for(Node n : graph[target]) {
					if(dist[n.target] > dist[target] + n.weight) {
						dist[n.target] = dist[target] + n.weight;
						pq.add(new Node(n.target, dist[n.target]));
					}
				}
			}
		}
		
		for(int i = 1; i <= v; i++) {
			if(dist[i] == Integer.MAX_VALUE)
				sb.append("INF\n");
			else
				sb.append(dist[i]).append("\n");
		}
		
		System.out.println(sb);
	}

}